### PR TITLE
snapstate: allow removal of snap.TypeOS when using a model with a base

### PR DESCRIPTION
**Pull request by mvo5**
This allows removal of the core snap (any snap.TypeOS) when using a system that uses a base in the model. This is a prerequisite to allow running the full spread suite on core18 ("reset" will remove all snaps including core by default there).

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -1297,6 +1297,15 @@ func canRemove(si *snap.Info, snapst *SnapState, removeAll bool) bool {
 		return true
 	}
 
+	// Allow snap.TypeOS removals if a different base is in use
+	if si.Type == snap.TypeOS {
+		if model, err := Model(st); err == nil {
+			if model.Base() != "" {
```
Comment (zyga) [inline overlord/snapstate/snapstate.go]: What if model.Base == si.Name? Is that captured by other code or is this check too simple?

Comment (mvo5) [reply overlord/snapstate/snapstate.go]: Thats a nice idea. My initial thinking was to ensure this via https://github.com/snapcore/snapd/pull/5358 but when reading this code people will probably wonder.

Comment (mvo5) [reply overlord/snapstate/snapstate.go]: I added a comment now that explains about boot-bases.

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1300,6 +1300,18 @@ func canRemove(si *snap.Info, snapst *SnapState, removeAll bool) bool {
 		return true
 	}
 
+	// Allow snap.TypeOS removals if a different base is in use
+	//
+	// Note that removal of the boot base itself is prevented
+	// via the snapst.Required flag that is set on firstboot.
```
Comment (zyga) [inline overlord/snapstate/snapstate.go]: Nice, thanks!

**Comment by pedronis** [comment]
we need at least a TODO about preventing removal of bases and core (as a base) if a snap  using them is present

**Pull request merged**